File: Moist.Application/Response.cs
namespace Moist.Application
{
    public static class Response
    {
        public static Response<Empty> Empty(bool success) => new Response<Empty>(new Empty(), "", !success);
        public static Response<T> Ok<T>(T data) => new Response<T>(data);
        public static Response<Empty> Ok(string message) => new Response<Empty>(new Empty(), message);
        public static Response<T> Ok<T>(T data, string message) => new Response<T>(data, message);
        public static Response<Empty> Fail(string message) => new Response<Empty>(new Empty(), message, true) ;
        public static Response<T> Fail<T>(string message) => new Response<T>(default, message, true) ;
        public static Response<T> Fail<T>(T data, string message) => new Response<T>(data, message, true);
    }
    public class Response<T>
    {
        public Response(T data, bool error = false)
            : this(data, "", error) {}

        public Response(T data, string message, bool error = false)
        {
            Data = data;
            Message = message;
            Error = error;
        }

        public T Data { get; }
        public bool Error { get; }
        public string Message { get; }
    }
}

File: Apps/moist_store/lib/models/Response.dart
import 'dart:convert';

import 'package:moist_store/models/Schema.dart';

import 'Shop.dart';

class Response<T>
{
  T data;
  String message;
  bool error;

  Response(this.data, this.message, this.error);

  factory Response.fromJson(dynamic json) {
    if(T == Shop){
      return Response(Shop.fromJson(json['data']) as T, json['message'] as String, json['error'] as bool);
    }
    return Response(null, json['message'] as String, json['error'] as bool);
  }
}

class CollectionResponse<T>
{
  List<T> data;
  String message;
  bool error;

  CollectionResponse(this.data, this.message, this.error);

  factory CollectionResponse.fromJson(dynamic json) {
    if(T == Schema){
      return CollectionResponse(_parseList<T>(json['data']), json['message'] as String, json['error'] as bool);
    }
    return CollectionResponse(null, json['message'] as String, json['error'] as bool);
  }

  static List<T> _parseList<T>(jsonData) {
    var list = jsonData as List;
    if(list.length == 0) return [];
    return list.map((x)=> Schema.fromJson(x)).toList() as List<T>;
  }

}
